src/sparseml/optim/helpers.py

```diff
@@ -512,3 +518,118 @@ def _maybe_parse_number(val: str) -> Union[str, float, int]:
             return val
+
+
+def _extract_metadata_from_recipe(container):
+    metadata = {}
+    if RECIPE_METADATA_KEY in container.keys():
+        metadata = container[RECIPE_METADATA_KEY]
+    return metadata
+
+
+def _extract_metadata_from_staged_recipe(container):
+    metadata = {}
+    stage_names = _get_recipe_stage_names(container)
+    for stage_name in stage_names:
+        if RECIPE_METADATA_KEY in container[stage_name].keys():
+            metadata[stage_name] = container[stage_name][RECIPE_METADATA_KEY]
+    if metadata and (len(metadata) != len(stage_names)):
+        raise ValueError(
+            "It seems that some stages in your checkpoint recipe"
+            "contain metadata and some don't."
+            "This is not allowed: either all the "
+            f"stages contain {RECIPE_METADATA_KEY} key or none does."
+        )
+
+    return metadata
+
+
+def _get_recipe_stage_names(container):
+    # Extracts valid stage names from a container.
+    # Valid stage name (key) is the one which corresponds to a value, which is
+    # a dictionary where at least one of the keys contains a string 'modifiers'.
+
+    stage_names = [
+        stage_name
+        for stage_name, stage_dict in container.items()
+        if isinstance(stage_dict, dict)
+        and any([key for key in stage_dict.keys() if "modifiers" in key])
+    ]
+    return stage_names
+
+
+def _check_warn_dict_difference(original_dict, new_dict):
+    if original_dict != new_dict:
+        warnings.warn(
+            f"Attempting to overwrite old metadata: {original_dict} "
+            f"with new metadata: {new_dict}. "
+            "It is prohibitive to alter the metadata of the recipe. "
+            "The new metadata will be omitted and discarded."
+        )
+
+
+def validate_metadata(metadata: dict, yaml_str: str) -> dict:
+    """
+    Compare the metadata (previous_metadata) carried over from the recipe
+    (`yaml_str`) with the new, incoming metadata ('metadata').
+
+    If attempting to overwrite previous metadata with the new metadata,
+    the script throws a warning and discards the new metadata.
+    Otherwise, it propagates the new metadata in the correct form.
+
+    :param metadata: New metadata
+    :param yaml_str: String representation of the recipe YAML file,
+        (may contain previous metadata)
+    :return: Validated metadata
+    """
+
+    container = load_recipe_yaml_str_no_classes(yaml_str)
+    is_container_staged = check_if_staged_recipe(container)
+
+    checkpoint_metadata = (
+        _extract_metadata_from_staged_recipe(container)
+        if is_container_staged
+        else _extract_metadata_from_recipe(container)
+    )
+
+    if checkpoint_metadata:
+        if metadata:
+            if is_container_staged:
+                for stage_name in _get_recipe_stage_names(container):
+                    # check if metadata is staged...
+                    if _get_recipe_stage_names(container) == list(metadata.keys()):
```

Comment:
can we move this check outside the for loop to see if the metadata is staged? Also would like to remove the ordering check (ie list equality vs set equality here) as it may not be enforced with plain dicts

Reply:
Good point. I would however place the `is_metadata_staged` check once we know that `checkpoint_metadata` and `metadata` and `is_container_staged`. Those are the logical required conditions before the `is_metadata_staged` check.